src/libstore/build/derivation-goal.cc

```diff
@@ -107,13 +113,7 @@ void DerivationGoal::killChild()
 void DerivationGoal::timedOut(Error && ex)
 {
     killChild();
-    done(BuildResult::TimedOut, {}, std::move(ex));
-}
-
-
-void DerivationGoal::work()
-{
-    (this->*state)();
+    auto _ = done(BuildResult::TimedOut, {}, std::move(ex));
```

Comment:
what is being returned?

Reply:
From what I can tell this shouldn't actually work because `amDone` is a coroutine and we never execute it.
The point was that `amDone` wouldn't be a coroutine but would return something you would await that never resumes,
i.e. you are in a coroutine, call `amDone`, you `co_await` it, and that would be the final suspension.

I will fix this, but I don't know why the tests pass. In my mind they shouldn't.

Reply:
In the end I fixed the bug I talked about above and this line is as it was before, except I added [[maybe_unused]] to silence the warning.

There is probably a better way of doing this, by splitting `done` into e.g. a function that returns `Done` and one that doesn't return `Done`, but I imagine the code will change heavily anyway in my next PR, so I refrained.

So to answer you question for what it does _now_:

`amDone` (called by `done`) does what it did before, but returns a `Done`, which has `[[nodiscard]]`.
This is done such that in a coroutine when you do this, it's clear that you should `co_return` it immediately (such that the coroutine ends, which is most often the behavior you want).
It doesn't do anything though, it's just a way of making the user do `co_return` and causing `final_suspend` to happen.
If you're not in a coroutine in the first place, obviously this makes no sense in the first place, hence we ignore the `Done` value.

Reply:
Thanks! That makes sense.